### ISSUE TITLE
Filter and sort functionality on par with dom-repeat

### ISSUE THREAD

**Issue by lukas-eibensteiner** [PolymerElements/iron-list#123]
It would be great if repeating elements like `iron-list` had `dom-repeat` functionality. So is there a way to sort and filter items in an iron-list while preserving bindings? Or did you explicitly decide against it?

I think this is another use-case for my [dom-repeat proposal](https://github.com/Polymer/polymer/issues/2183) in the main Polymer repository.

**Comment by blasten** [comment]
I agree that the list should support filtering and searching in the same way as dom-repeat.

**Comment by maria-le** [comment]
I also agree 100% we definitely need filtering and sorting capability for `iron-list`.

**Comment by capitanbatata** [comment]
In the meantime a workaround is suggested [in StackOverflow] (http://stackoverflow.com/questions/33139095/how-to-filter-an-iron-list-in-polymer-1-0).

**Comment by kasperstorgaard** [comment]
> In the meantime a workaround is suggested in StackOverflow.
For me, the main issue with the StackoverFlow computed filter workaround is that I lose selectedItem(s) when filtering like this.

**Comment by fakiolinho** [comment]
+1

**Comment by indolering** [comment]
The problem with the workaround list on StackOverflow is that iron-list resets it's scroll whenever you delete a record.

**Comment by tstewart15** [comment]
Is there a timeline on this feature? I don't see it on the Feature Backlog https://github.com/PolymerElements/iron-list/issues/10

**Comment by indolering** [comment]
So how would `iron-list` be able to discern between events such as a record being deleted and a filter search being performed?  In the former, you want `iron-list` to keep the scroll position but in the latter you want the scrollbar to be reset.

**Comment by indolering** [comment]
The canonical work-around for filter functionality has the unfortunate drawback of [resetting the list scroll position whenever the list is mutated](http://jsbin.com/dagiwa/2/edit?html,output).  I've created [an enhanced workaround](http://jsbin.com/lujufe/2/edit?html,output) which scrolls back to the correct list position.

**Comment by mherbert71** [comment]
Definitely something that is lacking in iron-list for sure.

**Comment by venicia** [comment]
Please consider this a higher priority, these functions should be built-in to this component.  The performance we gain by using this component is so needed for our applications but we lost the other important functions which are to provide the user the ability to filter and sort the list.

**Comment by sasivarnan** [comment]
Any update on this? :watch: :walking: :worried:

**Comment by indolering** [comment]
To everyone asking for updates, this is a complex feature and blasten is overworked!

@blasten maybe you could outline how you would like to have this implemented and ask others to contribute.

**Comment by blasten** [comment]
Hi all! Let me ask @kevinpschaaf. The only reference is dom-repeat in Polymer. https://github.com/Polymer/polymer/blob/master/src/lib/template/dom-repeat.html#L301-L309

**Comment by dtruffaut** [comment]
I :+1: this feature request.

The "observe" attribute, able to look for items subproperties changes (https://www.polymer-project.org/1.0/docs/devguide/templates#filtering-and-sorting-lists), is really needed for iron-list.

**Comment by 43081j** [comment]
If someone could have a go at this, that would be great. It should just be a copy (almost) of what dom-repeat does.

Being unable to sort is a huge problem, especially when dealing with firebase collections. Seeing as polymerfire splices items in one by one, any work around would result in triggering re-renders of the list per item.

**Comment by indolering** [comment]
@dtruffaut @43081j Please use reactions buttons and the star button to show your support, I don't want @blasten to have to lock this thread.

**Comment by 43081j** [comment]
One suggestion related to this, too, is that the functionality added for it be placed in a separate component.

Reason being, there are plenty of situations where we may want to sort/filter an array but maintain change notification/bindings. While I'm sure you could duplicate dom-repeat's logic to iron-list, it would not fix the problem for other situations needing this behaviour.

This would also mean dom-repeat could use the same element to handle this logic. Currently its very troublesome trying to implement this manually.

**Comment by tstewart15** [comment]
Hey @43081j , I actually already submitted an issue requesting this type of enhancement: https://github.com/Polymer/polymer/issues/3571

I created this component myself, and it works well, except when using it with iron-list, due to this Issue I filed againt iron-list: https://github.com/PolymerElements/iron-list/issues/300

**Comment by karthik-hande** [comment]
Even I am facing same issue while handling huge data

**Comment by phidias51** [comment]
It might be easier to break this issue into two separate issues: one for sorting and another for filtering.  The other thing to consider is that we already have specific array functions (this.splice, this pop, etc) to make it easier to manipulate arrays.  Perhaps this functionality should be added there, and then reused in array-based components like iron-list.

Filtering might be implemented by simply having the internal template test for validity. Something like this...

```html
  <iron-list ...>
    <template is="dom-if" test="_myFilterFunction">
     <paper-card ..></paper-card>
    </template>
  </iron-list>
```
This would obviate the need to change the model and maintain both a filtered and unfiltered model.

In general, how would sorting and filtering work in conjunction with firebase-query where results are returned not all at once but in drips and drabs? Would updates to the model trigger resorting/refiltering? 

If the user changes the filter function, the new filter should then be applied, and the list should be re-rendered. This commonly occurs when the user changes filtering criteria like "filter by company where company == 'Google'".
